gadget/raw.go

```diff
@@ -257,18 +260,78 @@ func (r *rawStructureUpdater) rollbackDifferent(out io.WriteSeeker, pc *LaidOutC
 	return nil
 }
 
+var eMMCDeviceRegex = regexp.MustCompile("mmcblk[0-9]boot[0-1]")
+
+// setEMMCPartitionReadWrite changes the read-write status of a eMMC boot partition.
+// Documentation for force_ro here:
+// https://www.kernel.org/doc/Documentation/mmc/mmc-dev-attrs.txt
+var setEMMCPartitionReadWrite = func(device string, rw bool) error {
+	sdevPath := fmt.Sprintf("/sys/block/%s/force_ro", device)
+
+	f, err := os.OpenFile(path.Join(dirs.GlobalRootDir, sdevPath), os.O_WRONLY, 0)
+	if err != nil {
+		return fmt.Errorf("cannot open device for writing: %v", err)
+	}
+	defer f.Close()
+
+	d := "1"
+	if rw {
+		d = "0"
+	}
+	if _, err := f.Write([]byte(d)); err != nil {
+		return fmt.Errorf("cannot change device read only mode to %v: %w", d, err)
+	}
+	return nil
+}
+
+func openDiskForWrite(device string, laidOutStruct *LaidOutStructure) (*os.File, func(), error) {
+	isEMMC := laidOutStruct.VolumeStructure.EnclosingVolume.Schema == schemaEMMC
+
+	var mmcDevice string
+	if isEMMC {
+		// get the emmc block name from the device path
+		mmcDevice = eMMCDeviceRegex.FindString(device)
+		if mmcDevice == "" {
+			return nil, nil, fmt.Errorf("%s is not a valid emmc block device", device)
+		}
+	}
+
+	setEMMCReadOnly := func() {
+		if isEMMC {
+			setEMMCPartitionReadWrite(mmcDevice, false)
+		}
+	}
+
+	if isEMMC {
+		if err := setEMMCPartitionReadWrite(mmcDevice, true); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	disk, err := os.OpenFile(device, os.O_WRONLY, 0)
+	if err != nil {
+		setEMMCReadOnly()
+		return nil, nil, fmt.Errorf("cannot open device for writing: %v", err)
+	}
+
+	return disk, func() {
+		disk.Close()
+		setEMMCReadOnly()
+	}, nil
```

Comment:
```suggestion
	return disk, func() error {
		return errors.Join(disk.Close(), setEMMCReadOnly())
	}, nil
```
but double check if our Go version allows for using errors.Join()

Reply:
Could not use errors.Join unfortunately, so went with other error checking method